# Learning Goals

In Assignment 1, we studied how information is represented by a single spiking neuron. In this assignment, you will learn how to construct networks of spiking neurons for a given cognitive task, how to propagate information through a network, and understanding the intuition behind network design choices. 

Let's first import all the libraries required for this assignment

In [ ]:
import math
import numpy as np
import matplotlib.pyplot as plt
import pickle

# Question 1: From single neuron to network of neurons [15 points]
## 1a.
What computational advantages do networks of neurons offer when compared against information processing by a single neuron? In other words, why do we need networks of neurons? 

## Answer 1a.

The network of neurons offer significant improvements over single neurons like  : 

**Distributed processing** :  Each individual neuron has a limited computational capability but when they work together in a network , each one can brings its unique perspective to a complex problem allowing for more accuracy.

**Adaptability** :  A single neuron's behavior is rigid and cannot be modified. In contrast network of neurons can adapt and better understand the environment by modifying connections(strengthening of good connections and weakening of bad connection aka updating weights).

**Parallel processing** : A network of neurons can process multiple inputs simultaneously enabling faster computation and handling of larger inputs, unlike a single neuron that can only process one input at a time.

**Non-linearity** : The combination of many nonlinear neurons can create complex nonlinear functions that can model complex systems.

**Robustness** : Networks of neurons are more robust to noise and errors in the input. A single neuron may be affected by small errors, but a network of neurons can still provide accurate results despite small errors.

**Fault tolerance** : Networks of neurons can continue to function even if some neurons fail. In contrast, a single neuron failure can lead to the complete failure of the system.


## 1b. 
Describe the algorithm for the information flow through a network of spiking leaky-integrate-and-fire (LIF) neurons. 

Specifically, trace out the steps required to compute network output from a given (continuous-valued) inputs (suppose it is the raw mnist image in assignment 1). 

The algorithm should describe  

Step (1) Encoding: how continuous-valued inputs are fed to the SNN input layer; 

Step (2) Processing: how the layer activations are computed; 

Step (3) Decoding: and how the output layer activity is decoded and used for downstream tasks (say if we want to use the output for some image classification task). 

Also, provide a diagrammatic overview of the algorithm to aid your explanation. 

You are free to assume any network size, and input and output dimensions. 

## Answer 1b.

**Encoding :**  

The continuous-valued inputs (such as the raw MNIST image in assignment 1) are first encoded into spike trains that can be processed by the networks. 
This involves mapping the input values to the firing rate of a group of input neurons, where the firing rate represents the intensity of the input.To achieve this, the input layer of the network consists of a set of neurons, each of which receives a particular input feature.The activity of each neuron in the input layer is determined by the corresponding feature value in the input vector.

For example, in the case of the MNIST image classification task, each pixel in the input image corresponds to an input feature, and the activity of each neuron in the input layer is proportional to the pixel value. This activity is then converted to a spike rate using a spike rate encoding function, such as the Poisson rate coding.

The input layer is typically connected to the first hidden layer through a fully connected weight matrix.

**Processing :**

The layer activations are computed by simulating the spiking behavior of the LIF neurons in the network. The LIF neuron model consists of three components: an input current, a membrane potential, and a spiking threshold. The input current is proportional to the firing rate of the input neurons, and it is integrated over time to update the membrane potential of the LIF neuron. Once the membrane potential reaches the spiking threshold, the LIF neuron generates a spike and its membrane potential is reset. The spike is then transmitted to the next layer of the network.

The activations of the hidden layer are computed by receiving and processing the spike trains from the input layer.

The neurons in the hidden layer integrate the incoming spikes over time, with each neuron having a specific set of synaptic weights that determine the strength of the connections between the input neurons and the hidden layer neurons.
Once the membrane potential of a neuron exceeds a threshold, the neuron spikes and sends an output signal to the next layer and its membrane potential is reset.

The weights of the connections between neurons determine the strength of the synaptic transmission, which affects the integration of incoming spikes.


**Decoding :**

The final step in the algorithm is to decode the output activity of the network to obtain a useful output signal. This involves mapping the spiking activity of the output neurons to a continuous-valued output signal.

In the case of the MNIST image classification task, the output layer of the network consists of a set of neurons, each of which corresponds to a particular digit class. The activity of each neuron in the output layer reflects the degree of confidence that the network has in assigning the input image to the corresponding class.

To obtain the final output, a decoding mechanism such as winner-takes-all or softmax is used to determine the most active neuron in the output layer, which corresponds to the predicted class label for the input image.

Diagramatic Overview

                    [ Continuous-valued inputs ]
                                |
                                V
                [ Input layer - Poisson spike train encoding ]
                                |
                                V
                       [ Spiking LIF neurons ]
                                |
                                V
                       [ Spike Propagation ]
                                |
                                V
                      [ Synaptic integration ]
                                |
                                V
                        [ Spike output ]
                                |
                                V
        [ Output spiking activity spike decoding winner-takes-all, softmax) ]
                                |
                                V
                        [ Final output ]

![image-2.png](attachment:image-2.png)

# Question 2: Elements of Constructing Feedforward Networks [20 points]
In this exercise, you will implement the two fundamental components of a feedforward spiking neural network: i) layers of neurons and ii) connections between those layers
## 2a. 
As the first step towards creating an SNN, we will create a class that defines a layer of LIF neurons. The layer object creates a collection of LIF neurons and applies input current to it (also called psp_input for postsynaptic input) to produce the collective spiking output of the layer. 

Below is the class definition for a layer of LIFNeurons. Fill in the components to define the layer. 

## Rubric: 10 points for correct implementation of class. 5 points for the right verification. [15 points]

In [ ]:
class LIFNeurons:
    """ Define Leaky Integrate-and-Fire Neuron Layer """

    def __init__(self, dimension, vdecay, vth):
        """
        Args:
            dimension (int): Number of LIF neurons in the layer
            vdecay (float): voltage decay of LIF neurons
            vth (float): voltage threshold of LIF neurons
        
        This function is complete. You do not need to do anything here.
        """
        self.dimension = dimension
        self.vdecay = vdecay
        self.vth = vth

        # Initialize LIF neuron states
        self.voltage = np.zeros(self.dimension)
#        self.current = np.zeros(self.dimension)
        self.spike = np.zeros(self.dimension)
#        self.cdecay = 0.5
#        self.vrest = 0
    
    def __call__(self, psp_input):
        """
        Args:
            psp_input (ndarray): synaptic input current at a single timestep. The shape of this is same as the number of neurons in the layer. 
        Return:
            self.spike: output spikes from the layer. The shape of this should be the same as the number of neurons in the layer. 
        
        Write the expressions for updating the voltage and generating the spikes for the layer given psp_input at one timestep. 
        """
        #Update the voltage
        #Generate the spikes from the voltage         
        #Reset the voltage if the neuron spikes

#         self.voltage[0] = psp_input[0]
#         if self.voltage[0] >= self.vth:
#           self.voltage[0] = self.vrest
#           self.spike[0] =1 

#         for t in range(1,self.dimension):    
#           self.current[t] = self.current[t-1]*self.cdecay + psp_input[t]
#           self.voltage[t] = self.voltage[t-1]*self.vdecay + self.current[t]
#           if self.voltage[t] >= self.vth:
#             self.voltage[t] = self.vrest
#             self.spike[t] =1 

        self.voltage = self.voltage * self.vdecay * (1 - self.spike) + psp_input
        self.spike = (self.voltage > self.vth).astype(float)
        
        return self.spike

To verify the correctness of your class implementation, create a layer of neurons using the class definition above, and pass through it random inputs. 

In [ ]:
#Create a layer of neurons using the class definition above
model_LIF = LIFNeurons( dimension = 20 ,vdecay = 0.5 , vth = 0.5)
#Create random input spikes with any probability and print them. Numpy random.choice function might be useful here. 
input_spikes = np.random.choice([0,1],20 , replace = True, p=[0.5,0.5])
print("Input Spikes : ", input_spikes)
#Propagate the random input spikes through the layer and print the output

output_spikes = model_LIF.__call__(input_spikes)
print("Output spikes : ",output_spikes )

Input Spikes :  [1 0 1 1 0 1 0 0 0 0 1 1 1 0 0 1 0 1 1 0]
Output spikes :  [1. 0. 1. 1. 0. 1. 0. 0. 0. 0. 1. 1. 1. 0. 0. 1. 0. 1. 1. 0.]


## 2b.
Now, we will create a class the defines the connection between a presynaptic layer and a postsynaptic layer. To create the connection, we need the activity of the presynaptic layer (also called presynaptic layer activation) and the weight matrix connecting the presynaptic and postsynaptic neurons. The output of the class should be the current for the postsynaptic layer. 

Below is the class definition for Connections. Fill in the components to create the connections. 

## Rubric: 3 points for correct implementation of class. 2 points for the right verification. [5 points]

In [ ]:
class Connections:
    """ Define connections between spiking neuron layers """

    def __init__(self, weights, pre_dimension, post_dimension):
        """
        Args:
            pre_dimension (int): number of neurons in the presynaptic layer
            post_dimension (int): number of neurons in the postsynaptic layer
            weights (ndarray): connection weights of shape post_dimension x pre_dimension

        This function is complete. You do not need to do anything here.

        """
        self.weights = weights
        self.pre_dimension = pre_dimension
        self.post_dimension = post_dimension
    
    def __call__(self, spike_input):
        """
        Args:
            spike_input (ndarray): spikes generated by the pre-synaptic neurons
        Return:
            psp: current for the post-synaptic neurons
        
        Write the operation for computing psp
        """
        
        #Compute psp given spike_input and self.weights
        psp = np.dot(self.weights, spike_input)
        
        return psp

To verify the correctness of your class implementation, create a connection object and compute the postsynaptic current for random presynaptic activation inputs and random connection weights. 

In [ ]:
#Define the dimensions of the presynaptic layer in a variable
pre_dimension = 100

#Define the dimensions of the postsynaptic layer in a variable
post_dimension = 50

#Create random presynaptic inputs with any probability. Numpy random choice function might be useful here. 
presynaptic_input_spike = np.random.choice([1,0], (pre_dimension,1), p=[0.6, 0.4])

#Create a random connection weight matrix. Numpy random rand function might be useful here. 
weight_matrix = np.random.rand(post_dimension, pre_dimension)

#Initialize a connection object using the Connection class definition and pass the variables created above as arguments
connection =  Connections(weight_matrix, pre_dimension, post_dimension)

#Compute the current for the postsynaptic layer when the connection object is fed random presynaptic activation inputs
psp = connection.__call__(presynaptic_input_spike)

#Print the shape of the current
print("current shape : ", psp.shape)


current shape :  (50, 1)


# Question 3: Constructing Feedforward SNN [15 points]
Now that you have implemented the basic elements of an SNN- layer and connection, you are all set to implement a fully functioning SNN. The SNN that you will implement here consists of an input layer, a hidden layer, and an output layer. 

Below is the class definition of an SNN. Your task is to create the layers and connections that form the network using the class definitions in Question 2. Then complete the function to propagate a given input through the network and decode network output. 

## Rubric: 10 points for correct implementation of class. 5 points for the right verification. [15 points]

In [ ]:
class SNN:
    """ Define a Spiking Neural Network with One Hidden Layer """

    def __init__(self, input_2_hidden_weight, hidden_2_output_weight, 
                 input_dimension=784, hidden_dimension=256, output_dimension=10,
                 vdecay=0.5, vth=0.5, snn_timestep=20):
        """
        Args:
            input_2_hidden_weight (ndarray): weights for connection between input and hidden layer. dimension should be hidden_dimension x input_dimension. 
            hidden_2_output_weight (ndarray): weights for connection between hidden and output layer. dimension should be output dimension x hidden dimension. 
            input_dimension (int): number of neurons in the input layer
            hidden_dimension (int): number of neurons in the hidden layer
            output_dimension (int): number of neurons in the output layer
            vdecay (float): voltage decay of LIF neurons
            vth (float): voltage threshold of LIF neurons
            snn_timestep (int): number of timesteps for simulating the network (also called inference timesteps)
        """
        self.snn_timestep = snn_timestep
        
        #Create the hidden layer
        self.hidden_layer = LIFNeurons(hidden_dimension, vdecay, vth)
        
        #Create the output layer
        self.output_layer = LIFNeurons(output_dimension, vdecay, vth)
        
        #Create the connection between input and hidden layer
        self.connection_input_to_hidden = Connections(input_2_hidden_weight,input_dimension,hidden_dimension )
        
        #Create the connection between hidden and output layer
        self.connection_hidden_to_output = Connections(hidden_2_output_weight,hidden_dimension,output_dimension)
        
    
    def __call__(self, spike_encoding):
        """
        Args:
            spike_encoding (ndarray): spike encoding of input
        Return:
            output: decoded output from the network
        """
        
        #Initialize an array to store the decoded network output for all neurons in the output layer
        spike_output = np.zeros(self.output_layer.dimension)
        
        #Loop through the simulation timesteps and process the input at each timestep tt
        for timestep in range(self.snn_timestep):
            
            #Propagate the input through the input to hidden layer and compute current for hidden layer
            hidden_current_in = self.connection_input_to_hidden.__call__(spike_encoding[timestep])
           
            #Compute hidden layer spikes 
            hidden_spikes = self.hidden_layer.__call__(hidden_current_in)
            
            #Propagate hidden layer inputs to output layer and compute current for output layer
            output_current_in = self.connection_hidden_to_output.__call__(hidden_spikes)
            
            #compute output layer spikes
            output_spikes = self.output_layer.__call__(output_current_in)
            
            #Decode spike outputs by summing them up
            spike_output += output_spikes
            
        return spike_output

To verify the correctness of your class implementation, define the arguments to initialize the SNN. Then initialize the SNN and pass through it random inputs and compute network outputs. 

In [ ]:
#Define the dimensions of the input layer in a variable
input_dimension=784

#Define the dimensions of the hidden layer in a variable
hidden_dimension = 256

#Define the dimensions of the output layer in a variable
output_dimension=10

#Define vdecay in a variable
vdecay=0.5

#Define vth in a variable
vth=0.5

#Define snn_timesteps in a variable
snn_timestep=20

#Create random input to hidden layer weights. Numpy random rand function might be useful here
input_2_hidden_weight = np.random.rand(hidden_dimension, input_dimension)


#Create random hidden to output layer weights. Numpy random rand function might be useful here
hidden_2_output_weight = np.random.rand(output_dimension, hidden_dimension)


#Create random spike inputs to the network. Numpy random choice function might be useful here
spike_input = np.random.choice([1,0], (snn_timestep,input_dimension), p=[0.6, 0.4])


#Print the inputs
print("Input Spikes of the nework  : \n" , spike_input)


#Create an SNN object using the class definition and variables defined above
snn = SNN(input_2_hidden_weight, hidden_2_output_weight, input_dimension, hidden_dimension, output_dimension,
                 vdecay, vth, snn_timestep)

#Pass the random spike inputs through the SNN and print the output of the SNN
output_spikes = snn.__call__(spike_input)
print("Output Spikes of the network : \n", output_spikes)

Input Spikes of the nework  : 
 [[0 1 1 ... 1 1 1]
 [1 0 0 ... 1 1 0]
 [1 1 1 ... 1 1 1]
 ...
 [1 0 1 ... 1 1 1]
 [0 1 1 ... 1 0 0]
 [0 0 0 ... 0 0 0]]
Output Spikes of the network : 
 [20. 20. 20. 20. 20. 20. 20. 20. 20. 20.]


# Question 4: SNN for Classification of Digits [25 points]
So far we have learnt how to construct SNNs for random inputs. In this exercise, you will use your implementation of SNNs to classify real-world data, taking the dataset of handwritten digits as an example. The dataset is provided as numpy arrays in the folder "data". Each sample in the MNIST dataset is a 28x28 image of a digit and a label (between 0 and 9) of that image. We will be dealing with batches, which means that we will read a fixed number of samples from the dataset (also called the batch size).

## 4a. 
First, we need to write two helper functions- to read the data from the saved data files, and to convert an image into spikes. The function to read the data is already written for you. You need to complete the function for encoding the data into spikes. 

## Rubric: 7 points for correct implementation of function. 3 points for the right verification. [10 points]

In [ ]:
def read_numpy_mnist_data(save_root, num_sample):
    """
    Read saved numpy MNIST data
    Args:
        save_root (str): path to the folder where the MNIST data is saved
        num_sample (int): number of samples to read
    Returns:
        image_list: list of MNIST image
        label_list: list of corresponding labels
    
    This function is complete. You do not need to do anything here.
    """
    image_list = np.zeros((num_sample, 28, 28))
    label_list = []
    for ii in range(num_sample):
        image_label = pickle.load(open(save_root + '/' + str(ii) + '.p', 'rb'))
        image_list[ii] = image_label[0]
        label_list.append(image_label[1])

    return image_list, label_list

def img_2_event_img(image, snn_timestep):
    """
    Transform image to spikes, also called an event image
    Args:
        image (ndarray): image of shape batch_size x 28 x 28
        snn_timestep (int): spike timestep
    Returns:
        event_image: event image- spike encoding of the image
        
    Complete the expression for converting the image to spikes (event image)
    """
    
    #Reshape the image. Do not touch this code
    batch_size = image.shape[0]
    image_size = image.shape[2]
    image = image.reshape(batch_size, image_size, image_size, 1)
    
    #Generate a random image of the shape batch_size x image_size x image_size x snn_timestep. Numpy random rand function will be useful here. 
    rand_img = np.random.rand(batch_size, image_size, image_size, snn_timestep)

    
    #Generate the event image
    event_image = (image > rand_img).astype(float)
    

    return event_image

To verify the correctness of your class implementation, load a sample digit from the saved file and convert it into an event image. Then print the shape of the event image. 

In [ ]:
#Load 1000 samples from the MNIST dataset using the read function defined above
images , labels = read_numpy_mnist_data("data/mnist_test",1000)


#Print the shape of the data
print("data shape : ",images.shape)


#Convert the images to event images
event_images = img_2_event_img(images , snn_timestep)


#Print the shape of the event image
print("Shape of the event image:- ",event_images.shape)



data shape :  (1000, 28, 28)
Shape of the event image:-  (1000, 28, 28, 20)


In [12]:

rand_img = np.random.rand(100, 28, 28, 20)

In [15]:
rand_img[0].reshape( -1 , rand_img.shape[-1]).shape

(784, 20)

## 4b. 
Next, we need another helper function to compute the classification accuracy of the network. The classification accuracy is defined as the percentage of the samples that the network classifies correctly. To compute the classification accuracy, you need to:

- Propagate each input through the network and obtain the network output.
- Based on the network output, the class of the image is the one for which the output neuron has maximum value. Let's call this predicted class. 
- Compare the predicted class against the true class. 
- Compute accuracy as the percentage of correct predictions. 

Below is the function for computing the test accuracy. The function takes in as arguments the SNN, directory in which the MNIST data is saved, and the number of samples to take from the MNIST dataset. Your task is to use the helper functions created above to load the data, convert into event images, and then compute network prediction and accuracies. 

## Rubric: 15 points for correct implementation of function

In [ ]:
def test_snn_with_mnist(network, data_save_dir, data_sample_num):
    """
    Test SNN with MNIST test data
    Args:
        network (SNN): defined SNN network
        data_save_dir (str): directory for the test data
        data_sample_num (int): number of test data examples
    """
    #Read image and labels using the read function
    test_image_list, test_label_list = read_numpy_mnist_data(data_save_dir, data_sample_num)
    
    #Convert the images to event images
    test_event_image_list = img_2_event_img(test_image_list, network.snn_timestep)
    
    
    #Initialize number of correct predictions to 0
    correct_prediction = 0
    
    #Loop through the test images
    for i in range(data_sample_num):
        #Compute network output for each image. You might have to reshape the image using Numpy reshape function so that its appropriate for the SNN
        output = test_event_image_list[i].reshape( -1 , test_event_image_list.shape[-1])
        output = output.T # making time steps as rows and the remaining as features
        network_output = network.__call__(output)
        
        #Determine the class of the image from the network output. Numpy argmax function might be useful here
        pred_class = np.argmax(network_output)
        
        
        #Compare the predicted class against true class and update correct_prediction counter
        if pred_class == test_label_list[i]:
            correct_prediction +=1
        
    #Compute test accuracy
    test_accuracy = (correct_prediction/data_sample_num)*100

    
    return test_accuracy

# Question 5: Tuning Membrane Properties for Correct Classification [25 points]
Great! We have everything that we need to measure the performance of the SNN for classification of MNIST digits. For this, we first need to create the SNN using the class definition we wrote in Q.3. Then we need to call the test function that we wrote in Q.4b. However, note that the SNN needs the connection weights between the layers as inputs. These weights are typically obtained as a result of "training" the network for a given task (such as MNIST classification). However, since training the network isn't a part of this assignment, we provide to you already trained weights. 

## 5a. 
Your task in this exercise is to initialize an SNN with vdecay=1.0 and vth=0.5. Test the SNN on MNIST dataset and obtain the classification accuracy.  

In [ ]:
#Load the weights. Do not touch this code
snn_param_dir = 'save_models/snn_bptt_mnist_train.p'
snn_param_dict = pickle.load(open(snn_param_dir, 'rb'))
input_2_hidden_weight = snn_param_dict['weight1']
hidden_2_output_weight = snn_param_dict['weight2']

#Define a variable for vdecay
vdecay = 1

#Define a variable for vth
vth = 0.5

#Create the SNN using the class definition in Q3 and the variables defined above
snn = SNN(input_2_hidden_weight, hidden_2_output_weight, input_dimension, hidden_dimension, output_dimension,
                 vdecay, vth, snn_timestep = 20)

#Compute test accuracy for the SNN on 1000 examples from MNIST dataset and print it
test_accuracy  = test_snn_with_mnist(snn ,"data/mnist_test" , 1000 )
print("Test Accuracy of MNIST : " , test_accuracy)



Test Accuracy of MNIST :  24.4


What could be a possible reason for the poor accuracy? (Hint: Consider the hyperparameters, e.g. the neuron properties, and also the parameter, e.g. the connections between neurons)

## Answer 5a. 

The Snn network model involes many hyperparameters like the voltage decay , current decay , resting potential , voltage threshold  etc . we need to tune these hyperparameters to identify the optimal values for attaining high accuracy



## Rubric: 7 points for correct training. 3 points for the explanation. [10 points]

## 5b. 
Can you tune the membrane properties (vdecay and vth) to obtain higher classification accuracies?

## Rubric: 10 points if accuracy above 90%. 8 points if between 70 and 90%. 5 points if between 50 and 70%. 0 points otherwise. 

In [ ]:
#Write your implementation of Question 5b. here

#Load the weights. Do not touch this code
snn_param_dir = 'save_models/snn_bptt_mnist_train.p'
snn_param_dict = pickle.load(open(snn_param_dir, 'rb'))
input_2_hidden_weight = snn_param_dict['weight1']
hidden_2_output_weight = snn_param_dict['weight2']

#Define a variable for vdecay
vdecay = 0.5

#Define a variable for vth
vth = 0.5

#Create the SNN using the class definition in Q3 and the variables defined above
snn = SNN(input_2_hidden_weight, hidden_2_output_weight, input_dimension, hidden_dimension, output_dimension,
                 vdecay, vth, snn_timestep = 20)

#Compute test accuracy for the SNN on 1000 examples from MNIST dataset and print it
test_accuracy  = test_snn_with_mnist(snn ,"data/mnist_test" , 1000 )
print("Test Accuracy of MNIST : " , test_accuracy)

Test Accuracy of MNIST :  97.3


## 5c.
Based on your response to Questions 5a. and 5b., can you explain how membrane properties affect network activity for classification?

## Answer 5c.

When the vdecay was high intially (Vdecay = 1) the accuracy was 24.4% but when vdecay reduced to 0.5 the accuracy significantly improved by a large margin attaining 97.3% .  


## Rubric: 3 points

## 5d.
Besides the membrane properties, what other impoertant factors can affect the classification accuracy? How should we improve it? Hint: Remember here we use a feedford network.

## Answer 5d.

1. Training data quality can be improved 
2. Make our model robust by training on images which are highly variational( noise in data )
3. Network architecture can be improved.( more hidden layers)
4. Regularization techniques on the network( something similar to dropout to prevent model overfitting)
5. Further tuning the Vdecay and Vth parameters

## Rubric: 2 points